WPMedia/fusion-news-theme-blocks
# Tmedia 456 offer paywall

## Update 10/15
The SubscriptionOverlay had a z-index issue where if I placed the Paywall block (that leverages the SubscriptionOverlay) lower in the PageBuilder hierarchy (for example in the main or footer section of one of our layouts), the browser’s internal z-index would give precedence to some components on the page (like the carousel) that show over the SubscriptionOverlay content.  My initial solution was to give a high z-index value to SubscriptionOverlay, but this had the side effect of also covering up the navigation block.  The other fix was to place the Paywall block as the first listed block in the Page Builder hierarchy.  But as Nick pointed out, this is a documentation/client nightmare.  So the solution is to leverage ReactDOM.createPortal .  By using this, I can place the content directly as a child of the DOM’s body element, therefore giving it a high internal z-index value, but without it covering up the nav block.  Also, it doesn’t matter where I place the Paywall block in the Page Builder in the hierarchy…I could even place it in the footer and it will work.  Takes a bit of a hack to allow Jest and Enzyme to run, but not too bad…I have documented the reasoning for it in the source code.

## Update 10/14
If you were previously testing this locally, I renamed the block from `Offer Paywall - Arc Block` to just`Paywall - Arc Block` as well as the actual feature name per Matthew's suggestion.  You may need to delete the .fusion directory in the feature pack before starting.

## Description
This add the Paywall overlay feature and accompanying component

## Jira Ticket
https://arcpublishing.atlassian.net/browse/TMEDIA-456

## Acceptance Criteria

- Paywall content component exists that renders paywall CONTENT (headline, subheadline, CTA, etc)
- Paywall Content component is rendered by dynamic paywall feature
- use offer hook to fetch offer details https://arcpublishing.atlassian.net/browse/TMEDIA-474
- Paywall renders offer headline & subheadline
- PB Editor can place a paid paywall on the page
- Paywall conditionally renders login link for anonymous users
- Paywall directs user to appropriate offer page
- paid paywall content renders if details are present

## Test Steps
- Add test steps a reviewer must complete to test this PR

1. Checkout this branch `git checkout tmedia-456-offer-paywall` Also get the latest from news-theme
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/identity-block,@wpmedia/subscriptions-block,@wpmedia/default-output-block`
3. In PB, find the page `demo-homepage` open the Publish version for the Gazette.  Then ensure you are logged out.
4. In PB find the page `kitchen-sink-article-jc` and open the `article-right-rail` version of the page in PB editor.
5. Add to the page the block `Paywall - Arc Block`, add the props and publish.
6. Then view the publish page for the Gazette and you should see the paywall.
7. Next go back to the published homepage and log in.
8. Go back to published `kitchen-sink-article-jc` and now you shouldn't see the paywall, cause your credit card was charged $350 for the annual subscription (sorry, there is no other way)

## Effect Of Changes
### Before
No Paywall period.


### After
Paywall is present if user is not logged in

## Dependencies or Side Effects
This branch is based of @nschubach branch (tmedia-454) that implements the usePaywall hook and output type changes

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x ] Confirmed all the test steps a reviewer will follow above are working.
- [x ] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [x ] Confirmed this PR has unit test files
  - [x ] Ran `npm run test`, made sure all tests are passing
  - [] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [ x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._
